draft-ietf-quic-load-balancers.md

```diff
@@ -991,12 +991,15 @@ Shared-State Retry Service Token {
 
 The tokens are protected using AES128-GCM as follows:
 
-* The token key and IV are retrieved using the Key Sequence.
+* Key Sequence is the 8 bit identifier to retrieve The token key and AEAD IV.
 
-* The nonce, N, is formed by combining the IV with the 96 bit unique token
-number. The 96 bits of the unique token number are left-padded with zeros to the
-size of the IV. The exclusive OR of the padded unique token number and the IV
-forms the AEAD nonce.
+* The AEAD IV, is a 96 bit sequence number maintained by configuration agent,
+and incremented by one after each key updating, its initial value can be an
+arbitrary random value.
+
+* The AEAD nonce, N, is formed by combining the AEAD IV with the 96 bit
+unique token number, the exclusive OR of the unique token number and the
+AEAD IV forms the AEAD nonce.
```

Comment:
This paragraph doesn't correctly capture what the original paragraph did. To generate the nonce:
1. left-pad the unique token number to the length of the IV
2. nonce = IV ^ (padded unique token number)

Reply:
Should we limit that length of nonce and IV is just 96 bit? See section 5.1 of RFC5116(https://tools.ietf.org/html/rfc5116#section-5.1)
```
AEAD_AES_128_GCM input and output lengths are as follows:

      K_LEN is 16 octets,

      P_MAX is 2^36 - 31 octets,

      A_MAX is 2^61 - 1 octets,

      N_MIN and N_MAX are both 12 octets, and

      C_MAX is 2^36 - 15 octets.
```

Reply:
I don't see much value in that. Selecting long keys and IVs probably has some security value that there isn't much reason to prohibit.

Reply:
Please have a look at this discussion: https://crypto.stackexchange.com/questions/41601/aes-gcm-recommended-iv-size-why-12-bytes

A 12 byte nonce(which was called IV in the link) is the best practice of  aes-gcm related cipher.

Further more, in cryptography related discussion, the definition of IV and nonce terms is always confusing, that's why I propose to add some explanation about our them in our draft

Reply:
It would be simpler to just add the paragraph describing the IV and keep the description of the nonce as is. Something like:
```
* The Key Sequence is the 8 bit identifier to retrieve the token key and IV.

* The AEAD IV, is a 96 bit data which produced by implementer's custom
AEAD IV derivation function.

* The AEAD nonce, N, is formed by combining the IV with the 96 bit unique token
number. The 96 bits of the unique token number are left-padded with zeros to the
size of the IV. The exclusive OR of the padded unique token number and the IV
forms the AEAD nonce.

Reply:
> It would be simpler to just add the paragraph describing the IV and keep the description of the nonce as is. Something like:
> 
> ```
> * The Key Sequence is the 8 bit identifier to retrieve the token key and IV.
> 
> * The AEAD IV, is a 96 bit data which produced by implementer's custom
> AEAD IV derivation function.
> 
> * The AEAD nonce, N, is formed by combining the IV with the 96 bit unique token
> number. The 96 bits of the unique token number are left-padded with zeros to the
> size of the IV. The exclusive OR of the padded unique token number and the IV
> forms the AEAD nonce.
> ```

Hi huitema:
Our current discussion is about whether we need to limit AEAD nonce to 96 bits, I find that 12-byte nonce is the best practice of aes-gcm related cipher, if we limit that, then we need not do padding for unique token number. Can you give us some better advice about it?

Reply:
In theory, the IV could be as large as the block size, 128 bits. In practice, it is indeed using 96 bits for AES GCM. @martinduke if we are OK with only supporting AES GCM, then we can fix the IV size to 96 bits.

Reply:
I guess that's where we are now. I'm tempted to have more crypto agility, but ultimately Retry Services are version dependent and everyone has to speak 128GCM in QUICv1 to handle Initial Packets. Let's use William's proposal.

Reply:
Actually, I take that back. Initial encryption doesn't protect anything and is much lower stakes. There should be a way to negotiate different ciphers.

William, let's get rid of the 12 Byte restriction. When this PR lands, I'll do another to introduce crypto agility.

Reply:
OK, I got it.